src/helpers.rs

```diff
@@ -50,114 +51,86 @@ pub trait TimedFuture: Sized + Future {
-				use chain::static_types;
-				use crate::chain::[<$runtime>]::runtime;
-
-				// maximum weight of the signed submission is exposed from metadata and MUST be this.
-				let max_weight = api
-					.constants()
-					.at(&runtime::constants().election_provider_multi_phase().miner_max_weight())
-					.expect("constant `miner_max_weight` must exist")
-					.ref_time;
-				let max_length = api
-					.constants()
-					.at(&runtime::constants().election_provider_multi_phase().miner_max_length())
-					.expect("constant `miner_max_length` must exist");
-				let max_votes_per_voter = api
-					.constants()
-					.at(&runtime::constants().election_provider_multi_phase().miner_max_votes_per_voter())
-					.expect("constant `miner_max_votes_per_voter` must exist");
-				let db_weight = api
-					.constants()
-					.at(&runtime::constants().system().db_weight())
-					.expect("constant `db_weight` must exist");
-
-				let system_db_weight =
-					frame_support::weights::RuntimeDbWeight { read: db_weight.read, write: db_weight.write };
+pub async fn mine_solution<T>(
+	api: &SubxtClient,
+	hash: Option<Hash>,
+	solver: Solver,
+) -> Result<(SolutionOf<T>, ElectionScore, SolutionOrSnapshotSize), Error>
+where
+	T: MinerConfig<AccountId = AccountId, MaxVotesPerVoter = static_types::MaxVotesPerVoter>
+		+ Send
+		+ Sync
+		+ 'static,
+	T::Solution: Send,
+{
+	let (voters, targets, desired_targets) = snapshot::<T>(&api, hash).await?;
+
+	let blocking_task = tokio::task::spawn_blocking(move || match solver {
+		Solver::SeqPhragmen { iterations } => {
+			BalanceIterations::set(iterations);
+			Miner::<T>::mine_solution_with_snapshot::<
+				SequentialPhragmen<AccountId, Accuracy, Balancing>,
+			>(voters, targets, desired_targets)
+		},
+		Solver::PhragMMS { iterations } => {
+			BalanceIterations::set(iterations);
+			Miner::<T>::mine_solution_with_snapshot::<PhragMMS<AccountId, Accuracy, Balancing>>(
+				voters,
+				targets,
+				desired_targets,
+			)
+		},
+	})
+	.await;
+
+	match blocking_task {
+		Ok(Ok(res)) => Ok(res),
+		Ok(Err(err)) => Err(Error::Other(format!("{:?}", err))),
+		Err(err) => Err(Error::Other(format!("{:?}", err))),
+	}
+}
 
-				static_types::DbWeight::set(system_db_weight);
-				static_types::MaxWeight::set(max_weight);
-				static_types::MaxLength::set(max_length);
-				static_types::MaxVotesPerVoter::set(max_votes_per_voter);
+pub async fn snapshot<T: MinerConfig + Send + Sync + 'static>(
+	api: &SubxtClient,
+	hash: Option<Hash>,
+) -> Result<Snapshot, Error>
+where
+	T: MinerConfig<AccountId = AccountId> + Send + Sync + 'static,
+	T::Solution: Send,
+{
+	let RoundSnapshot { voters, targets } = api
+		.storage()
+		.fetch(&runtime::storage().election_provider_multi_phase().snapshot(), hash)
+		.await?
+		.unwrap_or_default();
+
+	let desired_targets = api
+		.storage()
+		.fetch(&runtime::storage().election_provider_multi_phase().desired_targets(), hash)
+		.await?
+		.unwrap_or_default();
+
+	let voters: Vec<_> = voters
```

Comment:
I recall I once looked at why we need this hacky code and it seemed neccessary, I want to revisit it now again.

Reply:
it's to fetch to the snapshot data as we use `Miner::mine_solution_with_snapshot` added some docs.